components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/provider/RegistryBasedApplicationPermissionProvider.java

```diff
@@ -0,0 +1,314 @@
+            Collection permissionNode = tenantGovReg.newCollection();
+            permissionNode.setProperty("name", newName);
+            newApplicationNode = ApplicationMgtUtil.getApplicationPermissionPath() + PATH_CONSTANT + newName;
+            String applicationNode = newApplicationNode;
+            tenantGovReg.put(newApplicationNode, permissionNode);
+            addPermission(applicationNode, loadPermissions.toArray(new ApplicationPermission[loadPermissions.size()]),
+                    tenantGovReg);
+        } catch (RegistryException e) {
+            throw new IdentityApplicationManagementException(
+                    "Error while renaming permission node " + oldName + "to " + newName, e);
+        }
+    }
+
+    @Override
+    public void storePermissions(String applicationName, String username, PermissionsAndRoleConfig permissionsConfig)
+            throws IdentityApplicationManagementException {
+
+        int tenantId = MultitenantConstants.INVALID_TENANT_ID;
+        try {
+            tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+            IdentityTenantUtil.initializeRegistry(tenantId);
+        } catch (IdentityException e) {
+            throw new IdentityApplicationManagementException(
+                    "Error loading tenant registry for tenant domain: " + IdentityTenantUtil.getTenantDomain(tenantId),
+                    e);
+        }
+        Registry tenantGovReg = CarbonContext.getThreadLocalCarbonContext().getRegistry(RegistryType.USER_GOVERNANCE);
+
+        String permissionResourcePath = getApplicationPermissionPath();
+        try {
+            if (!tenantGovReg.resourceExists(permissionResourcePath)) {
+                boolean loggedInUserChanged = false;
+                UserRealm realm = (UserRealm) CarbonContext.getThreadLocalCarbonContext().getUserRealm();
+                if (!realm.getAuthorizationManager()
+                        .isUserAuthorized(username, permissionResourcePath, UserMgtConstants.EXECUTE_ACTION)) {
+                    // Logged-in user is not authorized to create the permission.
+                    // Temporarily change the user to the admin for creating the permission.
+                    PrivilegedCarbonContext.getThreadLocalCarbonContext()
+                            .setUsername(realm.getRealmConfiguration().getAdminUserName());
+                    tenantGovReg =
+                            CarbonContext.getThreadLocalCarbonContext().getRegistry(RegistryType.USER_GOVERNANCE);
+                    loggedInUserChanged = true;
+                }
+                Collection appRootNode = tenantGovReg.newCollection();
+                appRootNode.setProperty("name", "Applications");
+                tenantGovReg.put(permissionResourcePath, appRootNode);
+                if (loggedInUserChanged) {
+                    PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(username);
+                }
```

Comment:
this should be in the final block.

Reply:
Fixed: https://github.com/wso2/carbon-identity-framework/commit/a55181d5fdd11acf6f23808651ab50253ad85368